ocaml/xapi/repository_helpers.ml

```diff
@@ -283,24 +280,34 @@ module GuidanceSet = struct
     | l when eq_set2 l ->
         (* RestartDeviceModel and RestartToolstack *)
         ()
-    | l when eq_set3 l ->
-        (* RestartDeviceModel and EvacuateHost *)
-        ()
-    | l when eq_set4 l ->
-        (* EvacuateHost, RestartToolstack and RestartDeviceModel *)
-        ()
     | l ->
         let msg = error_msg l in
         raise Api_errors.(Server_error (internal_error, [msg]))
 
-  let resort_guidances ~kind gs =
-    match (find_opt RebootHost gs, kind) with
-    | Some _, _ ->
-        singleton RebootHost
-    | None, Recommended ->
+  let apply_precedence higher lowers gs =
+    match find_opt higher gs with
+    | Some _ ->
+        filter (fun elem -> not (List.mem elem lowers)) gs
+    | None ->
         gs
-    | None, Absolute ->
-        filter (fun g -> g <> EvacuateHost) gs
+
+  let precedences =
+    [
+      (RebootHost, [RestartToolstack; EvacuateHost; RestartDeviceModel])
+    ; (EvacuateHost, [RestartDeviceModel])
+    ]
+
+  let resort_guidances ~kind gs =
```

Comment:
I'm having a hard time understand this function - is it supposed to be equivalent to something like this? (I think it gets simpler if you use sets, so take `gs` to be a set below)

```
open GuidanceSet

let clean_guidances ~kind gs =
  let gs = 
    if mem RebootHost gs then
      gs |> remove RestartToolstack |> remove EvacuateHost |> remove RestartDeviceModel 
    else 
      gs
  in
  let gs = if mem EvauateHost gs then gs |> remove RestartDeviceModel else gs
  in match kind with
  | Recommended -> gs
  | Absolute -> remove EvacuateHost gs
```

Reply:
or use set difference instead of `remove .. |> remove .. |> remove ..`

Reply:
Using `Set` is a good idea. How about this:
```
let precedences =
  [
    (RebootHost, GuidanceSet.of_list [RestartToolstack; EvacuateHost; RestartDeviceModel])
  ; (EvacuateHost, GuidanceSet.of_list [RestartDeviceModel])
  ]

let resort_guidances ~kind gs =
  let gs' =
    List.fold_left
      (fun acc (higher, lowers) ->
        if GuidanceSet.mem higher gs then GuidanceSet.diff gs lowers
        else gs)
      gs precedences
  in
  match kind with
  | Recommended -> gs'
  | Absolute -> GuidanceSet.remove EvacuateHost gs'
```

I want to keep the `precedences` as I think this makes the precedence relation easy to be understood and updated in future.

Reply:
this looks good to me!